valhalla/filesystem_utils.h

```diff
@@ -0,0 +1,83 @@
+#include <sys/stat.h>
+
+#include <chrono>
+#include <filesystem>
+#include <fstream>
+#include <thread>
+
+#ifdef _WIN32
+#define FS_MTIME(st_stat) st_stat.st_mtime
+#elif __APPLE__
+#define FS_MTIME(st_stat) st_stat.st_mtime
+#else
+#define FS_MTIME(st_stat) st_stat.st_mtim.tv_sec
+#endif
+
+struct has_data_impl {
+  template <typename T, typename Data = decltype(std::declval<const T&>().data())>
+  static std::true_type test(int);
+  template <typename...> static std::false_type test(...);
+};
+
+template <typename T> struct has_data : decltype(has_data_impl::test<T>(0)) {};
+
+namespace std::filesystem {
+
+inline std::time_t last_write_time_t(const std::filesystem::path& p) {
+  // note, in C++20 there's a proper chrono::clock_cast so we can use filesystem::last_write_time
+  struct stat s;
+  if (stat(p.c_str(), &s) != 0)
+    throw std::runtime_error("could not stat " + p.string());
+  return FS_MTIME(s);
+}
+
+/**
+ * @brief Saves data to the path.
+ * @attention Will replace the contents in case if fpath already exists. Will create
+ * new directory if directory did not exist before hand.
+ * */
+template <typename Container>
+typename std::enable_if<has_data<Container>::value, bool>::type inline save(
+    const std::string& fpath,
+    const Container& data = {}) {
+  if (fpath.empty())
```

Comment:
this was quite messy code. I didn't clean up too much, could do a bit more. this was fairly painful to refactor

Reply:
did a bit more cleanup and changed the arg to `std::filesystem::path` to avoid a bunch of string allocations below